doc/design/refactor/session.md

```diff
@@ -0,0 +1,160 @@
+# Design Doc: Session
+
+## Abstract
+
+The *session* object encapsulates the environment in which the
+computation graph is executed.
+
+We will have *local* session and *remote* session, they offer the
+same [interface](#interface). The local session encapsulates the local
+runtime environment and the remote session encapsulates the cluster
+runtime envrionment.
+
+The local runtime envrionment contains:
+
+1. computation devices (i.e., CPU, GPU) handles, and
+1. the [scope](../scope.md) which holds all variables.
+
+The remote runtime envrionment contains:
+
+1. computation devices (i.e., CPU and GPU on node 0, 1) in a cluster,
+   and
+1. the distributed [scope](../scope.md) in a cluster which holds all
+   variables.
+
+The user can create a remote session on Paddle Cloud and evaluate the
+computation graph with it. In this way, the user can control the
+remote computation resource in a cluster from his local computer.
+
+
+## Background
+
+The current design has an implicit global session on which
+`paddle.eval()` is executed. The pain point is:
+
+Since the user is not able to explicitly switch between runtime
+environments such as the scope and the device contexts, the user
+cannot run a topology in two independent environments.
+
+For example, in reinforcement learning, the user may want to have a
+stale model for inference and a fresh model for training, and only
+replace the stale model with the fresh model periodically.
+
+Furthermore, we have no concept that encapsulates a remote environment
+that executes a computation graph.
+
+We need the session object to address above issues.
+
+
+## Session
+
+A session is an object that owns the runtime environment. All
+computations are executed through `session.eval`.
+
+
+### Interface
+
+```
+eval(
+    targets,
+    feed_dict=None,
+)
+```
+
+Evaluates the target Operations or Variables in `targets`.
+
+- *targets*: the evaluation targets. Can be a single Operation or
+  Variable, or a list with the Operations or Variables as elements.
+
+  The value returned by `eval()` has the same shape as the `target`
+  argument.
+
+  The computation graph is implicitly inferred from the targets.
+
+- *feed_dict*: a dictionary that contains the tensors which overrides
+  the edges of the computation graph.
+
+```
+close()
+```
+
+Closes the session. Calling this method releases the scope.
+
+
+### Create a Local Session
+
+```
+session(
+    gpu_ids=None
+)
+```
+
+Creates a new session. One session owns one scope, so creating
+multiple sessions will create different scopes.
+
+- *gpu_ids*: a single `int` or a list of `int` of the GPU IDs to be
```

Comment:
Setting up devices can take advantage of paddle v1 design, which default set `gpu_ids` to all of the available GPUs.

Can we use a string to specify devices, because there may be other devices than GPU, like FPGA, what TF does is `/job:worker/task:1/gpu:0` also can be `/job:worker/task:1/fpga:0`

Reply:
Great idea! Added.

Reply:
I thinking what's the difference between Session and [Executor](https://github.com/QiJune/Paddle/blob/9447858aa778bc4fe18419c746c2c6997d822633/doc/design/block_execution.md#run-time-analysis). The Executor has a Run interface to execute a ProgramDesc created in compile-time. DeviceContext(CPUDeviceContext/CUDADeviceContext) is created and managed by Executor.
And Maybe Executor is a data member of Session. Session will get a target and some device ids. The target is parsed to get a ProgramDesc. Then, the ProgramDesc and device ids are passed to Executor. Executor will created a DeviceContextManeger according to device ids. At last,  the ProgramDesc will be executed in specific hardwares.

Reply:
@QiJune Great! That is a clear logic. We could add one more step ProgramOptimizer (currently called Converter) between Session and Executor. Please see this graph for more detail: https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/refactor/distributed_architecture.md#local-training-architecture (In the graph, "PaddlePaddle runtime" means the Executor. Btw, there are many names of the same thing, we need to decide on the naming).

Reply:
After the offline discussion with @helinwang , I believe `Session` is 
- a manager of the resources, which is a more high-level concept than DeviceContextManager. It owns the resources.
- The job handler.
   - it's will construct a new graph according to the `targets` user given. 
   - Interact with remote cluster or local machine, fetch/feed tensor, start/stop running a graph....
   - The close interface allows user release the resource it owns.
The `Executor` is a runtime concept, which has nothing to do with Session.

Reply:
> it's will construct a new graph according to the targets user given.

The session will not "construct a new graph", it will send the graph to the converter, and the converter will construct a new graph.